### ISSUE TITLE
Navigation Controller feature request

### ISSUE THREAD

**Issue by BrendanSimon** [beeware/toga#785]
On iOS, to have native look and feel, a NavigationBar and NavigationControl needs to be implemented.  Common examples are the iOS Settings app and iOS Contacts app.

https://developer.apple.com/documentation/uikit/uinavigationcontroller

https://developer.apple.com/documentation/uikit/uinavigationbar

This needs to be abstracted to a common Toga API.  I presume Android has something similar.  Windows may have something similar with it's Windows Explorer address bar hierarchy.  Not sure about macOS or GTK (or wxPython or PyQt or PySide)

**Comment by freakboy3742** [comment]
Toga already has an NavigationView widget (which should possibly be called NavigationContainer; this would seem to be the match for the iOS UINavigationBar - a top level widget for navigating panes of content by "depth".

Closing this ticket on the basis there's already a toga.core API for this feature, and tickets aren't an especially effective way of coordinating efforts around specific platform implementations of widgets.

**Comment by BrendanSimon** [comment]
I can't see any mention of `NavigationView` (or `NavigationContainer`) in the docs.

https://toga.readthedocs.io/en/latest/reference/api/widgets/index.html

https://toga.readthedocs.io/en/latest/reference/api/index.html

https://toga.readthedocs.io/en/latest/reference/api/containers/index.html

**Comment by BrendanSimon** [comment]
Found it in the road map.

https://toga.readthedocs.io/en/latest/background/roadmap.html

I can see in the source code `NavigationView` where as the roadmap mentions `NavigationContainer`.

Can't see any examples for it, but there does seem to be some tests.

BTW, the `test_navigationview.py` has the following code.  Is it correct (the reference to "switch" looks suspicious to me)?

```python
    def test_widget_created(self):
        self.assertEqual(self.switch._impl.interface, self.switch)
        self.assertActionPerformed(self.switch, 'create Switch')
```

**Comment by freakboy3742** [comment]
Hrm... not sure what has happened there. The git history isn't being helpful, as the change predates the toga 0.3. restructure; I can only assume I stubbed out some code very early on, and because the widget doesn't work and isn't documented, nobody has been back to correct it.

However:
* The roadmap is correct - It should be `NavigationController`, not `NavigationView`.
* The absence from API docs is an oversight, probably driven by the fact that the only platform that makes even a cursory attempt at implementing NavigationView is iOS, and I know it doesn't work there.
* The reference to `switch` in the test almost certainly betrays the provenance of the code - it's a copy and paste error.

**Comment by deepio** [comment]
@freakboy3742 you can hit pypi api to get that information
https://pypistats.org/api/packages/toga/system
https://pypistats.org/api/packages/toga/python_minor

For more information about api endpoints, checkout:
https://pypistats.org/api/#api-packages-package-system

**Comment by freakboy3742** [comment]
Reopening on the basis that a navigation controller is a very important feature, especially for mobile apps, and despite the state of this ticket, we don't have one yet.

**Comment by johnzhou721** [comment]
Given that on iOS, the navigation bar is basically the window toolbar (see Apple HIG "Toolbar") which we're setting up in MainWindow logic, and same is true for macOS (see system preferences app, left and right arrows in toolbar).  On GTK4/libadwaita, the NavigationView widget also handles the window's top bar as seen at https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.7/class.NavigationView.html, so an abstraction for this functionality will mostly make sense on MainWindow.

### My proposal:

A new class, ``NavigationItem``, consisting of an (optional) title, and a widget.  2 functions, MainWindow.push_nav_item(item: NavigationItem) and MainWindow.pop_nav_item().  push_nav_item will push a new navigation item, i.e. navigate to a new widget / page, and pop_nav_item will remove the top navigation item.  The navigation item will also be popped when the back button is pressed.  Pushing a navigation item already in any MainWindows' stacks will raise an error.  NavigationItem will have 2 convenience properties:  1. visible, which is based on whether the navigation item is the currently at the top of a window, and 2. realized (for the lack of a better name), for if the navigation item is currently attached to a window.

By default, a window has 1 navigation item, the root navigation item.  This navigation item will represent what was previously represented by (MainWindow.content, MainWindow.title).  When attempting to pop this navigation item, an error will be raised.  The root navigation item will be stored in a property, MainWindow.root_navigation_item; however, for backwards compatibility, content and title can still be read and written, and it will be redirected to the root navigation item.  Reassigning the root navigation item itself pops all other navigation items automatically.  Attempting to pop the root item will result in an error.

There shall be a read-only property, navigation_items, on MainWindow, that is an array of all pushed navigation items so far on this window.  All references to MainWindow content shall be audited, and when, for example in the context of refresh, making references to the current view, shall be replaced by use of navigation_items.  There shall also be a convenience property, representing the top navigation item.  Setting it will be equivalent to pushing a new item.

**Comment by freakboy3742** [comment]
My immediate reaction: Does this need a new class? Isn't a "NavigationItem" a Container? (or even a Widget?) 

If the API is `MainWindow.push_content(title, widget)` and `MainWindow.pop_content()`, we'd be able to push any Widget onto the stack. `MainWindow.content` could then be a reference to the current top item of the content stack (possibly with a `root_content` as a universal reference to the top level widget). There might also be `MainWindow.content_stack` to give the full content stack.

It's also worth noting the potential overlap with #3445. Mobile-specific representations of Table and Tree will likely require integration with the navigation stack, as it's not plausible to show a full table/tree on mobile in the same way that you would on desktop. Each level of the stack won't be a true "Widget" for each layer of the stack (although it would certainly be possible to manufacture one as some sort of stateful wrapper around a parent widget).

**Comment by johnzhou721** [comment]
@freakboy3742 A class could facilitate us making certain things easier.  For example, we could have each NavigationItem maintain a list of toolbar items -- this is idiomatic on KDE, macOS, iOS, and probably GTK.  I don't think we should add toolbars as a first pass to avoid scope creep, but it's certainly something worth thinking about.

But then there might be more features we want to add, since each navigation item may manage the display of the window in some manner... so using a class can leave us open to that instead of changing the signature for the function each time.

> It's also worth noting the potential overlap with [#3445](https://github.com/beeware/toga/issues/3445). Mobile-specific representations of Table and Tree will likely require integration with the navigation stack, as it's not plausible to show a full table/tree on mobile in the same way that you would on desktop. Each level of the stack won't be a true "Widget" for each layer of the stack (although it would certainly be possible to manufacture one as some sort of stateful wrapper around a parent widget).

My 2c on this:  have the impls keep track of which navigations items are defined by interface, and when a navigation item must be pushed on the main window, pop all the ones that do not map to a widget interface, so if the user code pushes another navigation controller when the page for a Table/Tree is open, pop the page for Table/Tree.

Though -- this would preclude the user navigating to a different page inside some sort of thing in table/tree -- this is not existent right now, but being defensive would not hurt.

A better solution would perhaps be to, in some way, abstract the navigation presented by table/tree and group that with the navigation item that the table/tree belongs to.  So if you're popping things through code, both of these gets popped at the same time when we're at that layer.  And when those items are pushed again, we push the actual layer containing the table/tree only, not the layer instantiated by table/tree.

**Comment by mhsmith** [comment]
Related:
* https://github.com/beeware/toga/discussions/4059

**Comment by johnzhou721** [comment]
@mhsmith Thanks for spotting the overlap between this and tab bars; the tab bar discussion was actually incited because on iOS the current native hierarchy is UINavigationController > UIView > UITabBarController and this proved to be very painful during safe area updates and iOS 18-’s bar blurring effect, so we need to find a way to get UITabBarController > UINavigationController > UIview hierarchy.

So this naturally brings another issue: when there’s tabs, how do we manage navigation controllers?  iOS requires 1 navigation controller per tab with the tab bar constantly visible, but Android’s tabs are only at the root of the navigation structure, and stay on that page (so goes away as soon as you navigate to a different page).  I’ll think about how to abstract this efficiently when I get home tonight and make another post.

**Comment by freakboy3742** [comment]
> So this naturally brings another issue: when there’s tabs, how do we manage navigation controllers? 

I originally thought there would be a need for a Navigation "widget" (NavigationContainer) for exactly this reason. It matters whether an app has a top-level Option Container, with Tab N allowing deep navigation; or whether there's a deep navigation where the top-level widget has tabs. 

The NavigationContainer would then need to interact with the iOS NavigationController - in the same way that Tree and Table need to. But that becomes an internal implementation detail - iOS Widgets can push/pop context onto the internal navigation stack, but the external API is "a NavigationContainer that contains an OptionContainer that contains a Tree" (or whatever is required). I think this is essentially the "abstraction" you're referring to in your earlier reply.

Any questions about the stack of content are then questions for the API of the NavigationContainer - the *container* knows the stack of content it holds, the current top/bottom content, etc. 

There biggest complication would be the situation of a NavigationContainer containing a NavigationContainer (either directly, or indirectly - e.g. ,NC->OptionContainer->NC); but I don't think thats fundamentally a problem - it just requires some careful state management about what is currently on top of the stack.

**Comment by johnzhou721** [comment]
@freakboy3742 In pretty much all iOS native apps, use   NC > OptionContainer > NC will not be exactly acceptable, as this will mean that there are 2 navigation bars at the top.  No native apps do this.

But a Navigation "widget" doesn't really make sense IMHO, as there will be no navigation widgets that doesn't fill the full screen (esp. with iOS 26 updates, the animation will be really borked.)  Directly attaching to window to achieve correct hierachy seems like an *extremely* hacky approach to manage correctly to me, especially after #1969 already proves that directly attaching a widget to a window without a container *was* troublesome to maintain:

> Includes a Cocoa and iOS version of #1794. This allows us to remove the special case handling of the root content object in any given layout - constraints are used for all Toga widgets, rather than "all widgets except the root widget", and we now know that all widgets that are in a layout have a container.

(Some reasons at #4059 on whether to not use a widget also applies.  Sorry for circularly dependent comments.)

My proposal regarding this is to have a separate "navigation stack" for each tab (this depends on the plan I mentioned at https://github.com/beeware/toga/discussions/4059); on platforms that uses NavController > TabBar like Android, we display the root widgets of each navigation stack in the same root navigation item at the native level and code logic to switch tabs, and when pushing items we just push.

**Comment by freakboy3742** [comment]
> [@freakboy3742](https://github.com/freakboy3742) In pretty much all iOS native apps, use NC > OptionContainer > NC will not be exactly acceptable, as this will mean that there are 2 navigation bars at the top. No native apps do this.
> 
> But a Navigation "widget" doesn't really make sense IMHO, as there will be no navigation widgets that doesn't fill the full screen 

Only if you think of the widget as a container that *includes* the titlebar. I'm thinking of it as an organisational "sub region" in any given layout. 

For example, you have a main window that holds *multiple* navigation subwigets - say, "nav_a" and "nav_b". When you programmatically push "nav_a", you give the title and content for the new panel of content. That context can be popped, restoring the original mainwindow content. The content of nav_a could itself contain navigation sub contexts.

In practice, I agree it's likely that you'll only ever have a single navigation sub widget in a layout. It's also an example of a widget that might not actually "be" a widget, in the sense that it doesn't actually have any visual manifestation on it's own (i.e., you wouldn't set the background color of a NavigationContainer).

But - on iOS, a Tree or Table would effectively be a subclass of a NavigationContainer widget, because any given Tree defines a navigation context, and you could have multiple table widgets in a given layout.

**Comment by johnzhou721** [comment]
@freakboy3742 Interesting, and I like your approach to the abstraction... but the question here is if this helps keep track of things like this:

> I originally thought there would be a need for a Navigation "widget" (NavigationContainer) for exactly this reason. It matters whether an app has a top-level Option Container, with Tab N allowing deep navigation; or whether there's a deep navigation where the top-level widget has tabs.

So just to be clear, with your approach, a NavigationContainer will push or pop onto the currently "active" navigation stack, and the currently active navigation stack may depend on platform, because iOS requires a navigation controller per tab but Android requires an entire navigation controller around the tab interface?  So we could make "getting native navigation stack" down to a single endpoint which will make the abstraction for the widget easier.

Right now #4059 may be a dependency of this, because if we let OptionContainers auto-insert navigation bars, it only makes sense as part of top-level things and not if the optioncontainer is along with other things in the current layout.

**Comment by freakboy3742** [comment]
> [@freakboy3742](https://github.com/freakboy3742) Interesting, and I like your approach to the abstraction... but the question here is if this helps keep track of things like this:
> 
> > I originally thought there would be a need for a Navigation "widget" (NavigationContainer) for exactly this reason. It matters whether an app has a top-level Option Container, with Tab N allowing deep navigation; or whether there's a deep navigation where the top-level widget has tabs.
> 
> So just to be clear, with your approach, a NavigationContainer will push or pop onto the currently "active" navigation stack, and the currently active navigation stack may depend on platform, because iOS requires a navigation controller per tab but Android requires an entire navigation controller around the tab interface? So we could make "getting native navigation stack" down to a single endpoint which will make the abstraction for the widget easier.

I'm deliberately *not* making statements about implementation details at this point. We do need to maintain awareness of what is possible on a given platform, and the APIs that will be used to make that happen - but the first (and most important) step is to have a clear understanding of the external API that we're going to expose for this feature.

In this specific example, what I'm trying to highlight is that there's a difference between adding an NavigationContainer as a child of an OptionContainer tab, and adding an OptionContainer as a child of a NavigationContainer tab. 
 
> Right now [#4059](https://github.com/beeware/toga/discussions/4059) may be a dependency of this, because if we let OptionContainers auto-insert navigation bars, it only makes sense as part of top-level things and not if the optioncontainer is along with other things in the current layout.

I haven't yet got a coherent picture of what #4059 *is*, so that comparison seems somewhat premature.

**Comment by johnzhou721** [comment]
@freakboy3742 Last question first: #4059 is a discussion that I'm opening with the eventual goal to ask for a "tab content" option on a MainWindow itself (instead of a separate widget; the existing OptionContainer I hope will be used for specific sections of the app).  This is because it's top-level, and a widget wouldn't serve it well because if we make a "TopLevelOptionContainer", it's not guaranteed to be only 1 in a screen that fills the layout, and setting up the native hierachy will be difficult.

(Sidenote: Android also mostly requires the bottom bar to be always visible as well, when I looked at gmail it was an exception.)

> In this specific example, what I'm trying to highlight is that there's a difference between adding an NavigationContainer as a child of an OptionContainer tab, and adding an OptionContainer as a child of a NavigationContainer tab.

Ack; but as you said, let's get public APIs sorted first.  I'll make sure to consider this case.

---

So you would prefer a NavigationContainer widget that would push onto the currently active stack, if I'm right.  I'm okay with this part if you think it would simplify things; no comments here about this being a widget as long as we do careful state management, and it'll also make Tree etc easier.

For the public interface, the first thing is whether we'd want something like NavigationItem at all.  I'm still not convinced about saying no to this -- if we provide a variable to track all widgets and titles on the stack, for example, (Widget, title) seems a bit arbitary as a tuple if there's a "NavigationStack" or something like that, but also each navigation item you push also has some other properties, like its toolbar (and possibly layout behavior after #4039 comes up with a good proposal), that one may wants to modify.  So I think a class or even a NamedTuple will be good to keep track of the navigation item as something that you can push, instead of just a stray pair of properties.

Thoughts on this structure?  I'd very much appreciate feedback in this area before we get down to the specific naming of the public members of NavigationContainer.

[Also #4059 has overlap with this, but we could work out a solution that works for non-OptionContainer options first as long as we don't preclude letting it work properly later.  I'm comfortable discussing any specifics here but will stick to public API for now.]

**Comment by freakboy3742** [comment]
> So you would prefer a NavigationContainer widget that would push onto the currently active stack, if I'm right. I'm okay with this part if you think it would simplify things; no comments here about this being a widget as long as we do careful state management, and it'll also make Tree etc easier.

The reason I'm advocating for it is because it allows both `NavigationContainer(children=[OptionContainer())` and `OptionContainer(children=[NavigationContainer])`, and it's clear what both of those *do*. In the former, if you push the stack, the optionContainer tabs would *not* be visible; in the latter, they would be.

I'm open to other API suggestions; but any suggestion needs to either (a) be clear how this use case is satisfied, or (b) provide some sort of platform evidence that it's a use case that *can't* exist (and provide a high level explanation for *why* it isn't possible).
 
> For the public interface, the first thing is whether we'd want something like NavigationItem at all. I'm still not convinced about saying no to this -- if we provide a variable to track all widgets and titles on the stack, for example, (Widget, title) seems a bit arbitary as a tuple if there's a "NavigationStack" or something like that, but also each navigation item you push also has some other properties, like its toolbar (and possibly layout behavior after [#4039](https://github.com/beeware/toga/issues/4039) comes up with a good proposal), that one may wants to modify. So I think a class or even a NamedTuple will be good to keep track of the navigation item as something that you can push, instead of just a stray pair of properties.

My issue with NavigationItem isn't the [color of the bike shed](https://bikeshed.org). It's whether there actually *is* an entity that is a NavigationItem that needs to exist. 

What properties exist that need to be defined at the level of "Navigation"? Do you need to set a color? A label? My argument is that there *isn't* a "navigation item". `stack.push(NavigationItem(content, label)` and `stack.push(content, label)` become functionally identical, except that you don't need a whole other container class to manage the latter.

**Comment by johnzhou721** [comment]
> I'm open to other API suggestions; but any suggestion needs to either (a) be clear how this use case is satisfied, or (b) provide some sort of platform evidence that it's a use case that _can't_ exist (and provide a high level explanation for _why_ it isn't possible).
> 
> >

Even though I'm not proposing an alternative API, for (b):  At least on Apple platforms, HIG at https://developer.apple.com/design/human-interface-guidelines/tab-bars states:
> Make sure the tab bar is visible when people navigate to different sections of your app. If you hide the tab bar, people can forget which area of the app they’re in. The exception is when a modal view covers the tab bar, because a modal is temporary and self-contained.

So Navigation > Option can't be used, because navigating will hide the tab bar in the OptionContainer.  Also, Android guidelines at https://m3.material.io/components/navigation-bar/overview:

> Destinations don't change. They should be consistent across app screens.

Since "across app screens" is used, I'd assume that Navigation > Option wouldn't make sense, because for something to be consistent *across app screens*, it has to exist across app screens globally.

But I still like your NavigationContainer proposal, which we can put inside OptionContainers; each platform can then detect where the NavigationContainer is nested in and create the appropriate native hierarchy.  It also drops the dependency for #4059 or any related discussions to finish, since the APIs aren't so dependent on each other now that we decouple navigation into a widget.

---

> > For the public interface, the first thing is whether we'd want something like NavigationItem at all. I'm still not convinced about saying no to this -- if we provide a variable to track all widgets and titles on the stack, for example, (Widget, title) seems a bit arbitary as a tuple if there's a "NavigationStack" or something like that, but also each navigation item you push also has some other properties, like its toolbar (and possibly layout behavior after [#4039](https://github.com/beeware/toga/issues/4039) comes up with a good proposal), that one may wants to modify. So I think a class or even a NamedTuple will be good to keep track of the navigation item as something that you can push, instead of just a stray pair of properties.
> 
> My issue with NavigationItem isn't the [color of the bike shed](https://bikeshed.org). It's whether there actually _is_ an entity that is a NavigationItem that needs to exist.
> 
> What properties exist that need to be defined at the level of "Navigation"? Do you need to set a color? A label? My argument is that there _isn't_ a "navigation item". `stack.push(NavigationItem(content, label)` and `stack.push(content, label)` become functionally identical, except that you don't need a whole other container class to manage the latter.

On iOS, each native navigation item in the stack manages separate toolbar items.  I'm entirely okay with not including toolbar functionality as a first pass, and I think we can make forward compat work by using ``stack.push(content, label, toolbar=[])`` when we do add this functionality later.  Is this okay?

At a minimum our class for NavigationContainer will need a content property for the root content, obviously, a push(content, label) function, a pop() function that pops the top, a way to access the stack of (content, label)s, and perhaps some convenient properties to e.g. get top content, top label.  Do you think this would be enough for an initial implementation of this feature?  Thank you.

**Comment by freakboy3742** [comment]
> > I'm open to other API suggestions; but any suggestion needs to either (a) be clear how this use case is satisfied, or (b) provide some sort of platform evidence that it's a use case that _can't_ exist (and provide a high level explanation for _why_ it isn't possible).
> > >
> 
> Even though I'm not proposing an alternative API, for (b): At least on Apple platforms, HIG at https://developer.apple.com/design/human-interface-guidelines/tab-bars states:
...
> So Navigation > Option can't be used, because navigating will hide the tab bar in the OptionContainer. Also, Android guidelines at https://m3.material.io/components/navigation-bar/overview:

Ok - at this point I *maybe* see where the connection to #4059 fits. I've left some comments on that thread.

Assuming I have, that approach satisfies criteria (b) - we no longer have the "Nav->Option" / "Option->Nav" inconsistency, because lightweight option containers don't have top-level navigation implications; we only need to work out how to accomodate navigation. 
 
That likely also removes the need for a NavigationContainer - everything that would be "stack management" becomes an API on the main window (*not* Window, because Window doesn't have the titlebar and navigation controller to manage a stack of content).

> > What properties exist that need to be defined at the level of "Navigation"? Do you need to set a color? A label? My argument is that there _isn't_ a "navigation item". `stack.push(NavigationItem(content, label)` and `stack.push(content, label)` become functionally identical, except that you don't need a whole other container class to manage the latter.
> 
> On iOS, each native navigation item in the stack manages separate toolbar items. I'm entirely okay with not including toolbar functionality as a first pass, and I think we can make forward compat work by using `stack.push(content, label, toolbar=[])` when we do add this functionality later. Is this okay?
> 
> At a minimum our class for NavigationContainer will need a content property for the root content, obviously, a push(content, label) function, a pop() function that pops the top, a way to access the stack of (content, label)s, and perhaps some convenient properties to e.g. get top content, top label. Do you think this would be enough for an initial implementation of this feature? 

Ok - I guess I can see how having a NavigationItem that exists at a similar conceptual level as OptionItem would make sense - there's a label, and having an object that contains that label makes sense. I'm not wild about the idea of having lots of toolbar items... but that's also something I can see that people might want to do, and if they do, a NavigationItem would be the place to have that set up. So I guess having a wrapper would make sense.

If we don't need a standalone NavigationContainer, I guess we could put all this additional functionality on a `MainWindow.navigation` (or `MainWindow.nav`?) object. That object could be list-like - so:
* `MainWindow.nav[0]`returns the top level NavigationItem; 
* `MainWindow.nav[-1]` returns the currently displayed NavigationItem;
* (maybe?) `MainWindow.nav.current` is an alias for `MainWindow.nav[-1]`
* `MainWindow.nav[-1].title` returns/assigns the title of the currently displayed NavigationItem;
* `MainWindow.nav.push(NavigationItem)` pushes the navigation context (possibly also allowing `MainWindow.push(label, content=[], **kwargs)` is a shortcut for constructing a NavigationItem)
* `MainWindow.nav.pop()` pops the navigation context

Any thoughts on that as an API?

**Comment by johnzhou721** [comment]
@freakboy3742 Yes. I think this is a good API for doing this -- and should be clean to keep track of.  Accounting for #4059 though if this happens:

> We allow MainWindow content to accept the same inputs as OptionContainer.content (i.e., a list of OptionItems or equivalent).

though, we'll need nav for each content OptionItem.

**Comment by freakboy3742** [comment]
> [@freakboy3742](https://github.com/freakboy3742) Yes. I think this is a good API for doing this -- and should be clean to keep track of. Accounting for [#4059](https://github.com/beeware/toga/discussions/4059) though if this happens:
> 
> > We allow MainWindow content to accept the same inputs as OptionContainer.content (i.e., a list of OptionItems or equivalent).
> 
> though, we'll need nav for each content OptionItem.

True; I guess that poses question of whether `navigation` is a property of the *window*, or a property of the *content* of the window. 

The obvious API would be `window.navigation[1].push(...)` to pushing the context on the second top-level tab in the window; with `window.navigation[1][-1]` as a reference to the top level content of the second top-level tab. That is *conceptually* consistent, but it seems a little odd to me that you'd be indexing over different lists when dealing with navigation and tabs. Double indexing doesn't feel especially good either.

Putting the navigation on the OptionItem would mean `window.content[1].navigation.push(...)` - which is a little more explicit... but what do you do with the simple untapped case? `window.content.navigation.push()`? Plus - is it really the *content* that has navigation? If we do this - what is the value of `.navigation` on a button that is part of the layout of a page of content? `None`? `AttributeError`? A reference to the navigation context for the layout the widget is part of?

I guess we could split the difference and say that `Window.navigation` is an error if the content is tabbed; but that inconsistency doesn't sit right either. Although I guess if OptionItem has both content and navigation, it would be consistent with Window. 

Alternatively, `Window.navigation` could be made a shortcut for `Window.content[current_tab].navigation` in the tabbed case.

Any thoughts on these options?

**Comment by johnzhou721** [comment]
@freakboy3742 

> Putting the navigation on the OptionItem would mean `window.content[1].navigation.push(...)` - which is a little more explicit... but what do you do with the simple untapped case? `window.content.navigation.push()`? Plus - is it really the _content_ that has navigation? If we do this - what is the value of `.navigation` on a button that is part of the layout of a page of content? `None`? `AttributeError`? A reference to the navigation context for the layout the widget is part of?

I think the last item here will make some sense.  In implementing Table/Tree we can simply say self.interface.navigation.push instead of checking to see what context is appropriate here.

Not a huge fan of all other ideas, because we'd have to split further when we add Sidebar and potentially Sidebar > TabBar.

**Comment by freakboy3742** [comment]
> I think the last item here will make some sense. In implementing Table/Tree we can simply say self.interface.navigation.push instead of checking to see what context is appropriate here.

Agreed that's a potentially useful property; but it means every widget need to retain a concept of which navigation context it is a part of, which won't be easy to manage. 
 
> Not a huge fan of all other ideas, because we'd have to split further when we add Sidebar and potentially Sidebar > TabBar.

As with the discussion on #4059 - if you feel that Sidebar will materially impact on how navigation is implemented, we should possibly tackle that sooner rather than later.

**Comment by freakboy3742** [comment]
This feature request has been superseded by the discussion on #4271, and the specific DetailConatiner widget described by #4300.
